### PR TITLE
Add flow support

### PR DESCRIPTION
**Pull request by ryyppy**
See commit messages for more info

### CODE REVIEW

**Comment by ryyppy** [comment]
Solves issue #5

**Comment by zzarcon** [comment]
@oskarcieslik take a look ;)

**Comment by rpunkfu** [comment]
@zzarcon checking :)

File: gh-emoji.js [hunk 1]
```diff
@@ -13,19 +17,31 @@ export function load() {
...
 
-export function all() {
+export function all(): ?EmojiMap {
   return emojis;
 }
 
-export function exist(emojiId) {
-  return !!all()[emojiId];
+export function exist(emojiId: string): boolean {
+  const emojis = all(); // eslint-disable-line no-shadow
```
Comment (rpunkfu) [inline gh-emoji.js]: I don't think that we should disable eslint rules for specific blocks / lines, imo we should change the name of our global emojis variable to emojiMap or smth. What do you think?

File: gh-emoji.js [hunk 2]
```diff
@@ -13,19 +17,31 @@ export function load() {
...
+    return false;
+  }
+
+  return !!emojis[emojiId];
 }
 
-export function getUrl(emojiId) {
-  return all()[emojiId];
+export function getUrl(emojiId: string): ?string {
+  const emojis = all(); // eslint-disable-line no-shadow
```
Comment (rpunkfu) [inline gh-emoji.js]: Ditto here, look at comment above :)

File: types-test.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+/* @flow */
+
+import { load } from './dist/gh-emoji.js';
+
+
+// Flowtype should report this as an error
+const test: Promise<string> = load();
+
+test.then((result) => console.log(test)); //eslint-disable-line
```
Comment (rpunkfu) [inline types-test.js:9]: Same here, instead of turning eslint off maybe we should just remove unused argument?

**Comment by rpunkfu** [comment]
Hi @ryyppy! I checked this PR and my thoughts are:
- please remove all `"//eslint-disable-line"` comments from files and instead use solutions I proposed or if you have your own idea how to solve it, I'd love to hear it :)
- file `gh-emoji.js.flow` shouldn't go into `dist` folder since it's only for transpiled / minified .js code to be used in browser by putting it in tag. Instead please create folder `lib` / `src` and put it there. (Maybe we should create `flow` sub-folder for flow interfaces, declarations, etc?)
- I don't know what `types-test.js` is supposed to do, can you explain it to me? :)

Thanks @ryyppy :)

**Comment by ryyppy** [comment]
Cool, thanks for the instant review :-)

So for your points:
- I agree with `//eslint-disable-line` and will rename it to `emojiMap`
- I also agree for the `dist/gh-emoji.js.flow`... but if we set `dist/gh-emoji.js` as main, we need to make sure the `js.flow` file is in the same directory, otherwise `flowtype` won't pick up the definitions... I have two proposals how we could settle this:
  - Add a `cp src/gh-emoji.js.flow dist` in the dist script (probably the best solution for now)
  - Or put the declaration file as independent module declaration in the (newly official) flowtype repo (https://github.com/flowtype/flow-typed) - I will do that eventually, when this lib is stable
- `types-test.js` is basically a test file you can open, just to see if flowtype picks up the `js.flow` definitions correctly... this will be important for tests in the flow-typed project

Let me know what you think!

**Comment by zzarcon** [comment]
+1 for the `eslint` and for adding `cp src/gh-emoji.js.flow` into the `dist` npm script. 

Should be ready to merge after those changes :muscle:

**Comment by rpunkfu** [comment]
I think `cp src/gh-emoji.js.flow dist` is fine! :) How about having two directories `bin` and `dist`, where:
- `dist` will be for user who wants to manually download script and put it in script tag
- `bin` will be for npm's starting point

This way we can `cp src/gh-emoji.js.flow bin` and user will not be confused about `.flow` file in `dist`

**Comment by ryyppy** [comment]
By separating `js.flow` files into another directory, this will force flowtype users to add another `[lib]` path to their `.flowconfig`.

So they end up with another installation step (e.g. with following `.flowconfig`):

```
[lib]
interfaces/
node_modules/gh-emoji/bin/gh-emoji.js.flow
```

Do you think this is a huge problem for webdevs to deal with this extra files? Usually you also pack stuff into `dist` like `.js.map` or `.min.js` and they cherry pick stuff either via bower or via node? :-)

**Comment by rpunkfu** [comment]
Yeah, you're right :)

**Comment by ryyppy** [comment]
**To summarize and make this PR work I need to do following:**
- [x] Pull merge master (since functionality was added and this functionality is not typed yet)
- [x] Remove `//eslint-disable-line` stuff
- [x] Move the `flow.js` into a directory called `definitions`
- [x] Add the `cp definitions/gh-emoji.js.flow dist` to the`dist` command
- [ ] ~~Add the `flow check gh-emoji.js` to the `dist` command~~ (will postpone this for now)

**Comment by rpunkfu** [comment]
Yes, that's right :)

File: .eslintrc.js [hunk 4]
```diff
@@ -1,3 +1,8 @@
 module.exports = {
-  "extends": "airbnb/base"
-};
\ No newline at end of file
+  "parser": "babel-eslint",
+  "extends": "airbnb/base",
+  "plugins": [ "flow-vars" ],
+  "rules": {
+    "eqeqeq": [2, "smart"],
```
Comment (zzarcon) [inline .eslintrc.js:6]: ❤️

**Comment by zzarcon** [comment]
@ryyppy looks good for me 👍

**Comment by rpunkfu** [comment]
same here :)

**Pull request merged**